Fix: selection optimization in list disallow full list removal

sculpt0r opened the pull request:
<!--
🚨 If you want to submit a PR for a security vulnerability, please contact us directly
at https://ckeditor.com/contact/ instead. 🚨
-->
## What is the purpose of this pull request?

<!-- Bug fix / New feature / Typo fix / Other, please explain  -->

Bug fix

## Does your PR contain necessary tests?

All patches that change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [X] Unit tests
- [X] Manual tests

## Did you follow the CKEditor 4 code style guide?

Your code should follow the guidelines from the [CKEditor 4 code style guide](https://github.com/ckeditor/ckeditor4/blob/major/dev/docs/codestyle.md) which helps keep the entire codebase consistent.

- [X] PR is consistent with the code style guide

## What is the proposed changelog entry for this pull request?

```
* [#4931](https://github.com/ckeditor/ckeditor4/issues/4931): Fix: selection optimization skips list elements and disallow list removal.
```

## What changes did you make?

*Give an overview…*

Besides all details I put in the issue comments: I found that we have additional selection optimization which is responsible for shrinking selection under some circumstances. I added an additional check if the end selection container is a `li` element and prevents optimization if so.

## Which issues does your PR resolve?

Closes #4931 .
<!-- Closes #<ANOTHER_ISSUE_NUMBER>. -->

sculpt0r commented:
Rebased on the newest master.

KarolDawidziuk commented on core/selection/optimization.js:
```diff
@@ -117,15 +117,17 @@
 		}
 
-		if ( range.endOffset === 0 ) {
+		var isInList = range.endContainer.is && range.endContainer.is( 'li' );
+
+		if ( range.endOffset === 0 && !isInList )  {
```
Let's add a comment with a reference to the issue that this PR is solving and I think that short information about preventing shrinking selection in lists would be also nice.

```suggestion
		var isInList = range.endContainer.is && range.endContainer.is( 'li' );
        
        // Prevent shrink selection in lists (#4931).
		if ( range.endOffset === 0 && !isInList )  {
```

Also, is this additional check `range.endContainer.is` is needed? 🤔

sculpt0r replied:
I'll add a comment 👍 

The additional check is for the case when container is a text node - then it doesn't have `is` method which causes errors.

sculpt0r replied:
I slightly change the comment you proposed, since at this function level we don't know nth about 'shrinking' itself.

KarolDawidziuk commented on tests/core/selection/manual/optimizationlist.html:
```diff
@@ -0,0 +1,39 @@
+		</li>
+		<li>&nbsp;</li>
+	</ol>
+
```
Let's remove this unnecessary empty line. This comment applies also to the second list below. 
```suggestion
```

KarolDawidziuk commented on tests/core/selection/manual/optimizationlist.md:
```diff
@@ -0,0 +1,13 @@
+@bender-tags: selection, 4.19.2, bug, 4931
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, toolbar, elementspath, sourcearea, list, undo, div, table, image, basicstyles, format
```
Do we need to include this plugins: `div, table, image, format`?

KarolDawidziuk commented on tests/core/selection/manual/optimizationlist.md:
```diff
@@ -0,0 +1,13 @@
+@bender-ckeditor-plugins: wysiwygarea, toolbar, elementspath, sourcearea, list, undo, div, table, image, basicstyles, format
+
+
+1. Select entire content via `Ctrl+A`
```
It's a very small detail, but let's add dots at the end of each sentence.
```suggestion
1. Select entire content via `Ctrl+A`.
```

KarolDawidziuk reviewed (changes requested):
Hi @sculpt0r 

I've made an early review of your PR.

Overall fix working very well but unfortunately, it's not working on Firefox.
It's looking like after selecting the whole content by using `CTRL+A` `range.startContainter` and `range.endContainer` are the same and they contain the whole `body` instead of the first and last element from the list(line in the other browsers). 🤔  Could you look at it?

There are some also small things that can be improved. Please see my inline comments.

I think that we missed some unit tests to cover this issue.
If I remember correctly we currently have an existing test with removing lists so we can add it here. Please see https://github.com/ckeditor/ckeditor4/blob/t/4931/tests/plugins/list/list.js#L1
It would be good to add some cases with removing the whole mixed lists, lists with some basic styles etc.

sculpt0r commented:
Rebased on the newest master.

sculpt0r commented:
> Overall fix working very well but unfortunately, it's not working on Firefox.
It's looking like after selecting the whole content by using CTRL+A range.startContainter and range.endContainer are the same and they contain the whole body instead of the first and last element from the list(line in the other browsers). 🤔 Could you at it?

If you take a look at those containers during select all action on FF - they are always indicating `body` element. So it looks like some other bug with FF, but this time on 'delete' action. I changed the content in the manual test. It still proves this issue  (it is even closer to the content from the report). But it won't fail in FF. With the new content selection in FF also remains set on `body` - but this time the entire list is removed. Probably we should extract another issue for that - WDYT? In other case - that require a new 'debug' session to find out what cause this behaviour on FF.

> I think that we missed some unit tests to cover this issue.
If I remember correctly we currently have an existing tests with removing lists so we can add it here.

There are cases of 'removing' lists - but with 'deselecting' list feature - not with del/backspace key. Also, this issue is more about 'selection'. Because removing lists remains the same - the problem was that the selection was improper. So I would stick just to tests related to selection optimization.

KarolDawidziuk reviewed (approved):
Thanks, @sculpt0r for the improvements.

LGTM! 🎉 

I just add some style adjustments in the manual test by myself. 

> Probably we should extract another issue for that - WDYT? In other case - that require a new 'debug' session to find out what cause this behaviour on FF.

I think that we can extract it as a follow-up to see what's happened there. On the other hand, further delving into the issue may take some time and at this moment, we do not know how long it may take, so I'd like to hear the second reviewer's opinion about it.


#### It was an early review. In the next step, this PR will be checked again by a more experienced developer.

Comandeer reviewed (changes requested):
It seems that the fix works differently depending on the browser:

1.  In Firefox it works partially – the second editor in the manual test does not work (this was already discussed in the thread and the initial consensus is to move it to a follow-up).
2.  In Chrome the fix works correctly but only when selection is made via <kbd>Ctrl</kbd>+<kbd>A</kbd>. If the selection is made via our `selectall` plugin, Chrome starts behaving like Firefox: the last item in the second editor is not removed.
3.  In Safari the fix works correctly but something odd happens to the caret – it disappears after the deletion of the list.
4.  In IE8 and IE11 fix seems to work correctly.

Due to the fact that the issue in Firefox can also be reproduced in Chrome after using the `selectall` plugin, I would be for investigating it a little (even if only to check if the selection in these two cases is the same). It would be also good to check what happens in Safari and if we can fix it somehow.

Additionally, it be also good to add a unit test with a selection causing the issue (so selecting the list with the empty element at the end).

Please also see my inline comment about possible refactor.

sculpt0r commented:
> Due to the fact that the issue in Firefox can also be reproduced in Chrome after using the selectall plugin, I would be for investigating it a little (even if only to check if the selection in these two cases is the same).

Select all plugin in div editor in both browsers (FF & Chrome) makes selection range.startContainer and endContainer set on `div` with editor content:

![Screenshot 2022-08-23 at 09 14 42](https://user-images.githubusercontent.com/6498567/186095400-c63b25d0-0ae5-4a62-933e-fda2c7c0500e.png)
 
 So the selection optimization early returns with `false`:
 https://github.com/ckeditor/ckeditor4/blob/a7b0d54376a794f65bce64086bd5f6bb51cb02d9/core/selection/optimization.js#L116-L117
 
 Not removing a single list element is not a problem with selection optimization.

sculpt0r commented:
> In Safari the fix works correctly but something odd happens to the caret – it disappears after the deletion of the list.

The selection is set on the body after content deletion:
![image](https://user-images.githubusercontent.com/6498567/186102779-c55c77e5-6f76-4b07-b907-3ae9018c9b6b.png)

However, that happens only in this manual test 🤔 not happening in our dev samples:
 
![image](https://user-images.githubusercontent.com/6498567/186102485-770c49f0-3d67-49f2-b8da-30fa4d5973ac.png)

So the selection remains the same, but the carret is not visible 🤔

sculpt0r commented:
@Comandeer - ready for another review.

- I'm going to create a follow-ups
- Please see the above comments as a response to your recent review.
- Added automatic test
- Refactor code according to the suggestion
- Just in case (random manual test performing) added select all plugin to the manual test

Comandeer reviewed (changes requested):
@sculpt0r, could you check why the fix does not work with the `selectall` plugin? What's even more interesting, it seems to fail only for ordered lists, not unordered ones :thinking:

sculpt0r commented:
Rebased on the newest master

sculpt0r commented:
@Comandeer - as we spoke - please recheck the PR 🤞

sculpt0r commented:
Rebased on the newest master.

sculpt0r commented:
@Comandeer - as we spoke - I adjusted the manual test and reverted `ctrl+A` binding. Could you take a look at this PR again?

Comandeer commented:
Rebased onto the latest `master`.

Comandeer commented on tests/core/selection/manual/optimizationlist.md:
```diff
@@ -0,0 +1,14 @@
+@bender-tags: selection, 4.20.0, bug, 4931
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, toolbar, elementspath, sourcearea, list, undo, selectall
```
The `selectall` plugin does not seem to be used in the test.

```suggestion
@bender-ckeditor-plugins: wysiwygarea, toolbar, elementspath, sourcearea, list, undo
```

Comandeer commented on tests/core/selection/manual/optimizationlist.md:
```diff
@@ -0,0 +1,14 @@
+
+**Expected** The entire list content is removed from the editor.
+
+**Note** It is possible that list styling was not removed, so the first list item punctuation is in the content.
```
I'd rephrase this note a little because the results of unfixed code are very similar – the first list item marker is preserved. The only difference is the fact that unfixed code produces additional whitespace before it.

So maybe:

```suggestion
**Note** It is possible that list styling was not removed, so the first list item punctuation is in the content **without any preceding whitespace**.
```

Comandeer reviewed (changes requested):
The fix works fine. However, unit tests fail in IE8 – could you check it? Besides that, I have small remarks about the manual test.

sculpt0r commented:
Rebased on the newest master.

sculpt0r commented:
@Comandeer - I fixed the code for the IE8. It appears that the `endContainer` was not a `li` as in other browsers, but the text node inside the `li`. Because of that, the used method `is` does not exists on such node.

Also, adjusted the manual test. So the PR is ready for another review.

Comandeer commented:
Rebased onto the latest `master`.

Comandeer reviewed (approved):
LGTM!
